### PR TITLE
Baremetal 2.16.7 merge

### PR DESCRIPTION
**Pull request by AndrzejKurek**
Conflicts are described here: https://github.com/PelionIoT/pelion-crypto/commit/2eb27ec2e08c8ae4d06b1dc437dd58b4ca7fe53c
Additional (new) behavior added here: https://github.com/PelionIoT/pelion-crypto/commit/6ffed132c635aba25abdbcd70cf2ed08bc66966b

To view the conflicts, please locally run the steps described at the bottom of the page here:
https://github.com/PelionIoT/pelion-crypto/pull/18
After pressing "view command line instructions".
image

### CODE REVIEW

**Comment by AndrzejKurek** [comment]
Codesize change, checked by running ./scripts/baremetal.sh --rom --gcc:
libmbedcrypto.a: +0 Bytes
libmbedx509.a: +0 Bytes
libmbedtls.a: +44 Bytes
Total: +44 Bytes
No change in runtime heap size.

File: include/mbedtls/ssl.h [hunk 1]
```diff
@@ -143,6 +170,7 @@
 #define MBEDTLS_ERR_SSL_UNEXPECTED_CID                    -0x6000  /**< An encrypted DTLS-frame with an unexpected CID was received. */
 #define MBEDTLS_ERR_SSL_VERSION_MISMATCH                  -0x5F00  /**< An operation failed due to an unexpected version or configuration. */
 #define MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS                -0x7000  /**< A cryptographic operation is in progress. Try again later. */
+#define MBEDTLS_ERR_SSL_BAD_CONFIG                        -0x5E80  /**< Invalid value in SSL config */
```
Comment (shelib01) [inline include/mbedtls/ssl.h:173]: Only to me it's bothers that the errors jump with numbers ? :) :) :)

File: library/ecp.c [hunk 2]
```diff
@@ -118,6 +159,236 @@
 static unsigned long add_count, dbl_count, mul_count;
 #endif
 
+#if !defined(MBEDTLS_ECP_NO_INTERNAL_RNG)
+/*
+ * Currently ecp_mul() takes a RNG function as an argument, used for
+ * side-channel protection, but it can be NULL. The initial reasoning was
+ * that people will pass non-NULL RNG when they care about side-channels, but
+ * unfortunately we have some APIs that call ecp_mul() with a NULL RNG, with
+ * no opportunity for the user to do anything about it.
```
Comment (shelib01) [inline library/ecp.c:168]: They did the same as we did in tiny, just provided internal RNG for backward compatibility :)

File: library/ssl_cli.c [hunk 3]
```diff
@@ -179,10 +204,16 @@ static void ssl_write_signature_algorithms_ext( mbedtls_ssl_context *ssl,
...
     {
-        return;
+        return( 0 );
     }
 
-    MBEDTLS_SSL_DEBUG_MSG( 3, ( "client hello, adding signature_algorithms extension" ) );
+    MBEDTLS_SSL_DEBUG_MSG( 3,
+        ( "client hello, adding signature_algorithms extension" ) );
+
+#if !defined( MBEDTLS_SSL_END_FOR_EACH_SIG_HASH )
```
Comment (shelib01) [inline library/ssl_cli.c:213]: It seems the define should be MBEDTLS_SSL_CONF_SINGLE_SIG_HASH

File: library/ssl_cli.c [hunk 4]
```diff
@@ -194,11 +225,11 @@ static void ssl_write_signature_algorithms_ext( mbedtls_ssl_context *ssl,
...
-    if( end < p || (size_t)( end - p ) < sig_alg_len + 6 )
-    {
-        MBEDTLS_SSL_DEBUG_MSG( 1, ( "buffer too small" ) );
-        return;
-    }
+    /* Empty signature algorithms list, this is a configuration error. */
+    if( sig_alg_len == 0 )
+        return( MBEDTLS_ERR_SSL_BAD_CONFIG );
+
+    MBEDTLS_SSL_CHK_BUF_PTR( p, end, sig_alg_len + 6 );
```
Comment (shelib01) [inline library/ssl_cli.c:232]: Any known reason why the stopped to use MBEDTLS_SSL_CHK_BUF_PTR macro ?

File: library/ssl_cli.c [hunk 5]
```diff
@@ -258,27 +291,28 @@ static size_t ssl_get_ec_curve_list_length( mbedtls_ssl_context *ssl )
...
     elliptic_curve_len = 2 * ssl_get_ec_curve_list_length( ssl );
     if( elliptic_curve_len == 0 )
-        return;
+        return( MBEDTLS_ERR_SSL_BAD_CONFIG );
 
     if( end < p || (size_t)( end - p ) < 6 + elliptic_curve_len )
     {
         MBEDTLS_SSL_DEBUG_MSG( 1, ( "buffer too small" ) );
-        return;
+        return( MBEDTLS_ERR_SSL_BAD_CONFIG );
```
Comment (shelib01) [inline library/ssl_cli.c:315]: Why you didn't want to use the check buf macro in this place ?
 MBEDTLS_SSL_CHK_BUF_PTR( p, end, 6 + elliptic_curve_len );

File: library/ssl_cli.c [hunk 6]
```diff
@@ -291,23 +325,27 @@ static void ssl_write_supported_elliptic_curves_ext( mbedtls_ssl_context *ssl,
...
 
-    MBEDTLS_SSL_DEBUG_MSG( 3, ( "client hello, adding supported_point_formats extension" ) );
+    MBEDTLS_SSL_DEBUG_MSG( 3,
+        ( "client hello, adding supported_point_formats extension" ) );
 
     if( end < p || (size_t)( end - p ) < 6 )
     {
         MBEDTLS_SSL_DEBUG_MSG( 1, ( "buffer too small" ) );
-        return;
+        return( MBEDTLS_ERR_SSL_BAD_CONFIG );
```
Comment (shelib01) [inline library/ssl_cli.c:348]: Do we want to use macro in this case MBEDTLS_SSL_CHK_BUF_PTR( p, end, 6 ); ?

**Comment by shelib01** [comment]
Do you know why they removed MBEDTLS_SSL_CONF_FIXED_MAJOR_VER and MBEDTLS_SSL_CONF_FIXED_MINOR_VER  ?
ssl_cli.c lines 841 and 844

File: library/ssl_cli.c (hunk 5 above)
Comment (AndrzejKurek) [reply library/ssl_cli.c:315]: I didn't want to modify the code as much, and I wanted to stick to `MBEDTLS_ERR_SSL_BAD_CONFIG`. 
Using `MBEDTLS_SSL_CHK_BUF_PTR` would mean that there's no debug message (and maybe some tests depend on it), and the return code would be different (`MBEDTLS_ERR_SSL_BUFFER_TOO_SMALL`).

**Comment by AndrzejKurek** [comment]
> Do you know why they removed MBEDTLS_SSL_CONF_FIXED_MAJOR_VER and MBEDTLS_SSL_CONF_FIXED_MINOR_VER ?
> ssl_cli.c lines 841 and 844

It's added only in baremetal, by this commit:
https://github.com/PelionIoT/pelion-crypto/commit/381eaa59763d58ba4ab82acee4007b3771bd2d01

File: library/ssl_cli.c (hunk 4 above)
Comment (AndrzejKurek) [reply library/ssl_cli.c:232]: Do you mean they or us? 
It's introduced in mbedtls-2.16.7, so we weren't using it before.

**Comment by shelib01** [comment]
https://github.com/AndrzejKurek/pelion-crypto/blob/baremetal-2.16.7-merge/library/ssl_cli.c#L1271
Do you know why we didn't do "ssl->state++" as the original 2.16?

**Comment by shelib01** [comment]
Why in baremetal we dismiss this line "ssl->handshake->extended_ms = MBEDTLS_SSL_EXTENDED_MS_ENABLED" in this function?
https://github.com/PelionIoT/pelion-crypto/blob/f46790fa6e58a4928d55ecd9825c159e7db37471/library/ssl_cli.c#L1468

File: library/x509_crt.c [hunk 7]
```diff
@@ -855,6 +882,12 @@ static int x509_get_basic_constraints( unsigned char **p,
     if( *max_pathlen == INT_MAX )
         return( MBEDTLS_ERR_ASN1_INVALID_LENGTH );
 
+    /* Do not accept max_pathlen equal to INT_MAX to avoid a signed integer
+     * overflow, which is an undefined behavior. */
+    if( *max_pathlen == INT_MAX )
```
Comment (shelib01) [inline library/x509_crt.c]: Why do we need both checks line 882 and line 887 ?

File: tests/suites/test_suite_ecp.data [hunk 8]
```diff
@@ -336,6 +336,14 @@ ECP test vectors Curve448 (RFC 7748 6.2, after decodeUCoordinate)
 depends_on:MBEDTLS_ECP_DP_CURVE448_ENABLED
 ecp_test_vec_x:MBEDTLS_ECP_DP_CURVE448:"eb7298a5c0d8c29a1dab27f1a6826300917389449741a974f5bac9d98dc298d46555bce8bae89eeed400584bb046cf75579f51d125498f98":"a01fc432e5807f17530d1288da125b0cd453d941726436c8bbd9c5222c3da7fa639ce03db8d23b274a0721a1aed5227de6e3b731ccf7089b":"ad997351b6106f36b0d1091b929c4c37213e0d2b97e85ebb20c127691d0dad8f1d8175b0723745e639a3cb7044290b99e0e2a0c27a6a301c":"0936f37bc6c1bd07ae3dec7ab5dc06a73ca13242fb343efc72b9d82730b445f3d4b0bd077162a46dcfec6f9b590bfcbcf520cdb029a8b73e":"9d874a5137509a449ad5853040241c5236395435c36424fd560b0cb62b281d285275a740ce32a22dd1740f4aa9161cec95ccc61a18f4ff07"
 
+ECP point multiplication rng fail secp256r1
```
Comment (shelib01) [inline tests/suites/test_suite_ecp.data:339]: Do we have rng fail test for tiny ecc ?

File: include/mbedtls/ssl.h (hunk 1 above)
Comment (AndrzejKurek) [reply include/mbedtls/ssl.h:173]: Not only :D

**Comment by AndrzejKurek** [comment]
> Why in baremetal we dismiss this line "ssl->handshake->extended_ms = MBEDTLS_SSL_EXTENDED_MS_ENABLED" in this function?
> https://github.com/PelionIoT/pelion-crypto/blob/f46790fa6e58a4928d55ecd9825c159e7db37471/library/ssl_cli.c#L1468

It has to do with a rearrangement done by Hanno:
https://github.com/PelionIoT/pelion-crypto/commit/03b64fa6c12140793aaad7a25df906ad70182bda

File: library/x509_crt.c (hunk 7 above)
Comment (AndrzejKurek) [reply library/x509_crt.c]: Good catch. This fix was merged both into baremetal and the mainline mbedtls in a different form, and somehow git didn't pick this up as a conflict - just merged both of them. The old version should stay here, new one is dead code. I'll fix that!

File: tests/suites/test_suite_ecp.data (hunk 8 above)
Comment (AndrzejKurek) [reply tests/suites/test_suite_ecp.data:339]: Our random implementation can't fail. 
They have a mechanism for internal rng to fail in ECP here: https://github.com/PelionIoT/pelion-crypto/pull/19/files/fef1be1e86878ae15bf28641f59e3f27a457419e#diff-bc5fbd4345d02975f5698ac4ba2b16c767937a870392322aa3c66c3df58a8ee4R367

**Comment by AndrzejKurek** [comment]
CI is green, PR ready for re-review.

**Review by shelib01 — APPROVED** [review]

**Pull request merged**